File: BotsSDK/korebot/src/main/java/kore/botssdk/adapter/AdvanceListButtonAdapter.java
package kore.botssdk.adapter;

import static android.view.View.GONE;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.SharedPreferences;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.graphics.drawable.GradientDrawable;
import android.util.Base64;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.squareup.picasso.Picasso;

import java.util.ArrayList;

import kore.botssdk.R;
import kore.botssdk.listener.AdvanceButtonClickListner;
import kore.botssdk.listener.ComposeFooterInterface;
import kore.botssdk.listener.InvokeGenericWebViewInterface;
import kore.botssdk.models.BotResponse;
import kore.botssdk.models.Widget;
import kore.botssdk.net.SDKConfiguration;
import kore.botssdk.utils.BundleConstants;
import kore.botssdk.utils.StringUtils;
import kore.botssdk.viewUtils.RoundedCornersTransform;

@SuppressLint("UnknownNullness")
public class AdvanceListButtonAdapter extends RecyclerView.Adapter<AdvanceListButtonAdapter.ButtonViewHolder> {
    private final LayoutInflater inflater;
    private final ArrayList<Widget.Button> buttons;
    private String skillName;
    private final String type;
    private final AdvanceButtonClickListner advanceButtonClickListner;
    private AdvanceOptionsAdapter advanceOptionsAdapter;
    private final ComposeFooterInterface composeFooterInterface;
    private InvokeGenericWebViewInterface invokeGenericWebViewInterface;
    private int count;
    private final boolean isOptions;
    private final SharedPreferences sharedPreferences;

    public AdvanceListButtonAdapter(Context context, ArrayList<Widget.Button> buttons, String type, AdvanceButtonClickListner advanceButtonClickListner, ComposeFooterInterface composeFooterInterface, InvokeGenericWebViewInterface invokeGenericWebViewInterface, boolean isOptions) {
        this.buttons = buttons;
        this.inflater = LayoutInflater.from(context);
        this.type = type;
        this.isOptions = isOptions;
        this.advanceButtonClickListner = advanceButtonClickListner;
        this.composeFooterInterface = composeFooterInterface;
        this.invokeGenericWebViewInterface = invokeGenericWebViewInterface;
        sharedPreferences = context.getSharedPreferences(BotResponse.THEME_NAME, Context.MODE_PRIVATE);
    }

    @NonNull
    @Override
    public AdvanceListButtonAdapter.ButtonViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        if (!StringUtils.isNullOrEmpty(type) && type.equalsIgnoreCase(BundleConstants.FULL_WIDTH)) {
            return new ButtonViewHolder(inflater.inflate(R.layout.advance_button_fullwidth, viewGroup, false));
        }
        return new ButtonViewHolder(inflater.inflate(R.layout.widget_button_list_item, viewGroup, false));
    }

    @Override
    public void onBindViewHolder(@NonNull AdvanceListButtonAdapter.ButtonViewHolder holder, int position) {
        Widget.Button btn = buttons.get(position);
        holder.tv.setText(btn.getTitle());
        holder.ivBtnImage.setVisibility(GONE);
        if (isOptions) {
            GradientDrawable gradientDrawable = (GradientDrawable) (holder.layoutDetails.getBackground() != null ? holder.layoutDetails : holder.buttonsLayout).getBackground().mutate();
            if (position % 2 == 0) {
                gradientDrawable.setColor(Color.parseColor(SDKConfiguration.BubbleColors.quickReplyColor));
                gradientDrawable.setStroke(1, Color.parseColor(SDKConfiguration.BubbleColors.quickReplyColor));
                holder.tv.setTextColor(Color.parseColor(SDKConfiguration.BubbleColors.quickReplyTextColor));
            } else {
                String bgColor = sharedPreferences.getString(BotResponse.BUBBLE_LEFT_BG_COLOR, "#ffffff");
                String txtColor = sharedPreferences.getString(BotResponse.BUBBLE_LEFT_TEXT_COLOR, "#000000");
                gradientDrawable.setColor(Color.WHITE);
                gradientDrawable.setStroke(2, Color.parseColor(bgColor));
                holder.tv.setTextColor(Color.parseColor(txtColor));
            }
        } else {
            holder.tv.setTextColor(Color.parseColor(SDKConfiguration.BubbleColors.quickReplyColor));
            ViewGroup parent = (ViewGroup) holder.tv.getParent();
            parent.setBackgroundResource(R.drawable.list_button_background);
        }

        if (!StringUtils.isNullOrEmpty(btn.getIcon())) {
            holder.ivBtnImage.setVisibility(View.VISIBLE);
            try {
                String imageData;
                imageData = btn.getIcon();
                if (imageData.contains(",")) {
                    imageData = imageData.substring(imageData.indexOf(",") + 1);
                    byte[] decodedString = Base64.decode(imageData.getBytes(), Base64.DEFAULT);
                    Bitmap decodedByte = BitmapFactory.decodeByteArray(decodedString, 0, decodedString.length);
                    holder.ivBtnImage.setImageBitmap(decodedByte);
                } else {
                    Picasso.get().load(btn.getIcon()).transform(new RoundedCornersTransform()).into(holder.ivBtnImage);
                }
            } catch (Exception e) {
                holder.ivBtnImage.setVisibility(GONE);
            }
        }

        holder.tv.setOnClickListener(view -> {
            if (!StringUtils.isNullOrEmpty(btn.getBtnType())) {
                if (btn.getBtnType().equalsIgnoreCase("confirm")) {
                    if (advanceOptionsAdapter != null)
                        advanceButtonClickListner.advanceButtonClick(advanceOptionsAdapter.getData());
                } else {
                    if (advanceOptionsAdapter != null)
                        advanceOptionsAdapter.setChecked(-1);
                }
            } else {
                if (composeFooterInterface != null && invokeGenericWebViewInterface != null) {
                    if (BundleConstants.BUTTON_TYPE_URL.equalsIgnoreCase(btn.getType())) {
                        invokeGenericWebViewInterface.invokeGenericWebView(btn.getUrl());
                    } else {
                        String title = btn.getTitle();
                        String payload = btn.getPayload();
                        composeFooterInterface.onSendClick(title, payload, false);
                    }
                }
            }

            if (advanceButtonClickListner != null)
                advanceButtonClickListner.closeWindow();
        });

        holder.layoutDetails.setOnClickListener(view -> {

            if (!StringUtils.isNullOrEmpty(btn.getBtnType())) {
                if (btn.getBtnType().equalsIgnoreCase("confirm")) {
                    if (advanceOptionsAdapter != null)
                        advanceButtonClickListner.advanceButtonClick(advanceOptionsAdapter.getData());
                } else {
                    if (advanceOptionsAdapter != null)
                        advanceOptionsAdapter.setChecked(-1);
                }
            } else {
                if (composeFooterInterface != null && invokeGenericWebViewInterface != null) {
                    if (BundleConstants.BUTTON_TYPE_URL.equalsIgnoreCase(btn.getType())) {
                        invokeGenericWebViewInterface.invokeGenericWebView(btn.getUrl());
                    } else {
                        String title = btn.getTitle();
                        String payload = btn.getPayload();
                        composeFooterInterface.onSendClick(title, payload, false);
                    }
                }
            }

            if (advanceButtonClickListner != null)
                advanceButtonClickListner.closeWindow();
        });
    }

    public void setInvokeGenericWebViewInterface(InvokeGenericWebViewInterface invokeGenericWebViewInterface) {
        this.invokeGenericWebViewInterface = invokeGenericWebViewInterface;
    }

    @Override
    public int getItemCount() {
        if (count != 0)
            return count;

        return buttons != null ? buttons.size() : 0;
    }

    public void setDisplayLimit(int count) {
        this.count = count;
    }

    public static class ButtonViewHolder extends RecyclerView.ViewHolder {
        final TextView tv;
        final ImageView ivBtnImage;
        final LinearLayout layoutDetails;
        final LinearLayout buttonsLayout;

        public ButtonViewHolder(@NonNull View itemView) {
            super(itemView);
            tv = itemView.findViewById(R.id.buttonTV);
            ivBtnImage = itemView.findViewById(R.id.ivBtnImage);
            layoutDetails = itemView.findViewById(R.id.layout_deails);
            buttonsLayout = itemView.findViewById(R.id.buttonsLayout);
        }
    }

    public void setListAdapter(AdvanceOptionsAdapter listAdapter) {
        this.advanceOptionsAdapter = listAdapter;
    }

    public String getSkillName() {
        return skillName;
    }

    public void setSkillName(String skillName) {
        this.skillName = skillName;
    }
}

File: BotSdk/ui/src/main/java/com/kore/ui/adapters/AdvanceListButtonAdapter.kt
package com.kore.ui.adapters

import android.content.Context
import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.graphics.toColorInt
import androidx.recyclerview.widget.RecyclerView
import com.kore.data.repository.preference.PreferenceRepositoryImpl
import com.kore.event.BotChatEvent
import com.kore.model.constants.BotResponseConstants
import com.kore.model.constants.BotResponseConstants.BUTTON_TYPE_URL
import com.kore.model.constants.BotResponseConstants.THEME_NAME
import com.kore.ui.R
import com.kore.ui.utils.BitmapUtils

class AdvanceListButtonAdapter(
    val context: Context,
    val type: String,
    private val buttons: List<Map<String, *>>,
    private val count: Int,
    private val isLastItem: Boolean,
    private val isOptions: Boolean,
    val actionEvent: (event: BotChatEvent) -> Unit
) : RecyclerView.Adapter<ButtonViewHolder>() {
    private var advancedOptionsAdapter: AdvanceOptionsAdapter? = null
    private val sharedPreferences = PreferenceRepositoryImpl()

    fun setAdvancedOptionsAdapter(advancedOptionsAdapter: AdvanceOptionsAdapter?) {
        this.advancedOptionsAdapter = advancedOptionsAdapter
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ButtonViewHolder {
        if (type.isNotEmpty() && type == BotResponseConstants.FULL_WIDTH) {
            return ButtonViewHolder(LayoutInflater.from(context).inflate(R.layout.advance_button_full_width, parent, false))
        }

        return ButtonViewHolder(LayoutInflater.from(context).inflate(R.layout.widget_button_list_item, parent, false))
    }

    override fun getItemCount(): Int {
        return when (count > 0) {
            true -> count
            else -> buttons.size
        }
    }

    override fun onBindViewHolder(holder: ButtonViewHolder, position: Int) {
        val button = buttons[position]

        holder.buttonTV.text = button[BotResponseConstants.KEY_TITLE] as String
        holder.ivBtnImage.visibility = View.GONE

        if (button[BotResponseConstants.ICON] != null) {
            holder.ivBtnImage.visibility = View.VISIBLE
            try {
                val imageData: String = button[BotResponseConstants.ICON] as String
                BitmapUtils.loadImage(context, imageData, holder.ivBtnImage, R.drawable.ic_image_photo)
            } catch (e: Exception) {
                holder.ivBtnImage.visibility = View.GONE
            }
        }
//        val txtColor = sharedPrefs.getStringValue(holder.itemView.context, THEME_NAME, BotResponseConstants.BUBBLE_RIGHT_BG_COLOR, "#3F51B5")
//        holder.buttonTV.setTextColor(txtColor.toColorInt())
        if (isOptions) {
            val gradientDrawable = (holder.buttonTV.parent as ViewGroup).background.mutate() as GradientDrawable
            if (position % 2 == 0) {
                val bgColor: String = sharedPreferences.getStringValue(context, THEME_NAME, BotResponseConstants.BUBBLE_RIGHT_BG_COLOR, "#ffffff")
                val txtColor: String = sharedPreferences.getStringValue(context, THEME_NAME, BotResponseConstants.BUBBLE_RIGHT_TEXT_COLOR, "#000000")
                gradientDrawable.setColor(bgColor.toColorInt())
                gradientDrawable.setStroke(1, bgColor.toColorInt())
                holder.buttonTV.setTextColor(txtColor.toColorInt())
            } else {
                val bgColor: String = sharedPreferences.getStringValue(context, THEME_NAME, BotResponseConstants.BUBBLE_LEFT_BG_COLOR, "#ffffff")
                val txtColor: String = sharedPreferences.getStringValue(context, THEME_NAME, BotResponseConstants.BUBBLE_LEFT_TEXT_COLOR, "#000000")
                gradientDrawable.setColor(Color.WHITE)
                gradientDrawable.setStroke(2, bgColor.toColorInt())
                holder.buttonTV.setTextColor(txtColor.toColorInt())
            }
        } else {
            val txtColor: String = sharedPreferences.getStringValue(context, THEME_NAME, BotResponseConstants.BUBBLE_RIGHT_BG_COLOR, "#3F51B5")
            holder.buttonTV.setTextColor(txtColor.toColorInt())
//            val parent = holder.tv.getParent() as ViewGroup
//            parent.setBackgroundResource(R.drawable.list_button_background)
        }
        holder.buttonTV.setOnClickListener {
            advancedOptionsAdapter?.let {
                if (button[BotResponseConstants.BTN_TYPE] != null) {
                    if (!isLastItem) return@setOnClickListener
                    if (button[BotResponseConstants.BTN_TYPE] == BotResponseConstants.CONFIRM) {
                        if (it.getSelectedItems().isNotEmpty())
                            actionEvent(BotChatEvent.SendMessage("Confirm: ${it.getSelectedItems().joinToString(",")}"))
                    } else {
                        it.clearSelectedItems()
                    }
                } else {
                    if (button[BotResponseConstants.TYPE] == BUTTON_TYPE_URL) {
                        actionEvent(BotChatEvent.UrlClick(button[BotResponseConstants.URL].toString()))
                    } else {
                        if (!isLastItem) return@setOnClickListener
                        val title: String = button[BotResponseConstants.KEY_TITLE].toString()
                        val payload: String = button[BotResponseConstants.PAYLOAD].toString()
                        actionEvent(BotChatEvent.SendMessage(title, payload))
                    }
                }
            }
        }

        holder.layoutDetails.setOnClickListener {
            val listElementButtonPayload = if (button[BotResponseConstants.PAYLOAD] != null) {
                button[BotResponseConstants.PAYLOAD] as String
            } else ""

            val listElementButtonTitle = if (button[BotResponseConstants.KEY_TITLE] != null) {
                button[BotResponseConstants.KEY_TITLE] as String
            } else ""

            try {
                when (button[BotResponseConstants.TYPE]) {
                    BotResponseConstants.USER_INTENT,
                    BotResponseConstants.URL,
                    BotResponseConstants.WEB_URL -> {
                        if (button[BotResponseConstants.URL] != null) {
                            actionEvent(BotChatEvent.UrlClick(button[BotResponseConstants.URL] as String))
                        }
                    }

                    else -> {
                        if (!isLastItem) return@setOnClickListener
                        actionEvent(BotChatEvent.SendMessage(listElementButtonTitle, listElementButtonPayload))
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }
}


class ButtonViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
    val buttonTV: TextView = itemView.findViewById(R.id.buttonTV)
    val ivBtnImage: ImageView = itemView.findViewById(R.id.ivBtnImage)
    val layoutDetails: LinearLayout = itemView.findViewById(R.id.layout_details)
}
